### PR TITLE
Refine cudnn softmax

### PR DESCRIPTION
**Pull request by GaoWei8**
### PR types
Performance optimization

### PR changes
OPs

### Describe
增加softmax cudnn的 channel 模式，替换原有cudnn代码。

### softmax前向

|api |  Paddle(kernel)优化前 | Paddle(kernel)优化后 | Tensorflow(kernel) | 参数|
|---|---|---|---|---|
|softamx_5 |--|**0.18842**|	0.60431|input(Variable)-dtype:float32,shape:[128,128,16,16] axis(int):0 use_cudnn(bool):True|

###  softmax反向

|api |  Paddle(kernel)优化前 | Paddle(kernel)优化后 | Tensorflow(kernel) | 参数|
|---|---|---|---|---|
|softamx_5 |--|**0.3256**|	0.82062	|	input(Variable)-dtype:float32,shape:[128,128,16,16] axis(int):0 use_cudnn(bool):True|

### CODE REVIEW

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 1]
```diff
@@ -14,58 +14,100 @@ limitations under the License. */
...
+    auto* Out = ctx.Output<Tensor>("Out");
 
     auto dims = X->dims();
-    auto flattened_dims = framework::flatten_to_2d(dims, dims.size() - 1);
-    framework::LoDTensor flattened_x;
-    framework::LoDTensor flattened_out;
-    flattened_x.ShareDataWith(*X).Resize(flattened_dims);
-    flattened_out.ShareDataWith(*Out).Resize(flattened_dims);
-
-    math::SoftmaxCUDNNFunctor<T>()(
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc:40]: 不再调用`SoftmaxCUDNNFunctor`的话，`SoftmaxCUDNNFunctor`这个函数在别处应该没有调用了，可以删除。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 2]
```diff
@@ -14,58 +14,100 @@ limitations under the License. */
...
+        desc_, CUDNN_TENSOR_NCHW, type, N, axis_dim, D, 1));
+
+    auto& dev_ctx = ctx.template device_context<platform::CUDADeviceContext>();
+    auto handle = dev_ctx.cudnn_handle();
+    auto mode = axis == rank - 1 ? CUDNN_SOFTMAX_MODE_INSTANCE
+                                 : CUDNN_SOFTMAX_MODE_CHANNEL;
+
+    PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::cudnnSoftmaxForward(
+        handle, CUDNN_SOFTMAX_ACCURATE, mode,
+        platform::CudnnDataType<T>::kOne(), desc_, X->data<T>(),
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: https://github.com/PaddlePaddle/Paddle/blob/12bed4a9312d0df120c27ace65028283d918f247/paddle/fluid/operators/math/softmax.cu#L48-L56

原实现中使用了2个desc，这里只用1个，会有影响吗？

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 3]
```diff
@@ -14,58 +14,100 @@ limitations under the License. */
...
+
+    auto* out_data = Out->data<T>();
+    cudnnTensorDescriptor_t desc_;
+    cudnnDataType_t type = platform::ToCudnnDataType(
+        framework::ToDataType(std::type_index(typeid(T))));
+
+    PADDLE_ENFORCE_CUDA_SUCCESS(
+        platform::dynload::cudnnCreateTensorDescriptor(&desc_));
+    PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::cudnnSetTensor4dDescriptor(
+        desc_, CUDNN_TENSOR_NCHW, type, N, axis_dim, D, 1));
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: - softmax支持5-D的，单测里面构造下5-D的测试用例。
- 若输入是NCHW，对最后一个维度（W）做softmax，这里构造的cudnn Tensor的维度是{NCH, W, 1, 1}？

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 1 above)
Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc:40]: 在 softmax_with_cross_entropy_op.cu文件中有调用这个函数计算。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 3 above)
Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: - 已添加。

- 是的，维度是{NCH, W, 1, 1}。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 2 above)
Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: 之前发现相同格式的cudnnTensorDescriptor_t，部分是可以共用的。

**Comment by GaoWei8** [comment]
> 若是axis对应最后一个维度，那调用的softmax实现应该是一样的吧？为什么这个PR会快呢？

初步测试是因为cudnnTensorDescriptor_t的定义方式不同造成的性能差别。后期考虑找类似用法测试下。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 4]
```diff
@@ -12,60 +12,98 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+    const int axis = CanonicalAxis(ctx.Attr<int>("axis"), rank);
+    const int axis_dim = dims[axis];
+    const int N = SizeToAxis(axis, dims);
+    const int D = SizeOutAxis(axis, dims);
+
+    cudnnTensorDescriptor_t desc_;
+    cudnnDataType_t type = platform::ToCudnnDataType(
+        framework::ToDataType(std::type_index(typeid(T))));
+
+    PADDLE_ENFORCE_CUDA_SUCCESS(
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: 可以用ScopedTensorDescriptor。你这里Create了，但是忘了Destroy了。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 5]
```diff
@@ -12,60 +12,98 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-  void Compute(const framework::ExecutionContext& context) const override {
-    auto* X = context.Input<Tensor>("X");
-    auto* Out = context.Output<Tensor>("Out");
-
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* X = ctx.Input<Tensor>("X");
+    auto* Out = ctx.Output<Tensor>("Out");
     // allocate memory on device.
-    Out->mutable_data<T>(context.GetPlace());
+    Out->mutable_data<T>(ctx.GetPlace());
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: 变量名不符合命名规范。

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 6]
```diff
@@ -12,60 +12,98 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-    framework::LoDTensor flattened_out;
-    flattened_x.ShareDataWith(*X).Resize(flattened_dims);
-    flattened_out.ShareDataWith(*Out).Resize(flattened_dims);
-
-    math::SoftmaxCUDNNFunctor<T>()(
-        context.template device_context<platform::CUDADeviceContext>(),
-        &flattened_x, &flattened_out);
+    const int rank = dims.size();
+    const int axis = CanonicalAxis(ctx.Attr<int>("axis"), rank);
+    const int axis_dim = dims[axis];
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc]: `axis_dim`是什么意思，不要堆积名词。

Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: Done.

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 4 above)
Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: Done.

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 5 above)
Comment (GaoWei8) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc]: Done.

**Review by luotao1 — APPROVED** [review]
LGTM
